core/lib/src/rocket.rs

```diff
@@ -792,10 +787,45 @@ impl Rocket {
             .executor(runtime.executor())
             .serve(service);
 
-        // TODO.async: Use with_graceful_shutdown, and let launch() return a Result<(), Error>
-        runtime.block_on(server).expect("TODO.async handle error");
+        let (future, handle) = server.remote_handle();
+        runtime.spawn(future);
+        Ok(handle)
+    }
+
+    /// Starts the application server and begins listening for and dispatching
+    /// requests to mounted routes and catchers. Unless there is an error, this
+    /// function does not return and blocks until program termination.
+    ///
+    /// # Error
+    ///
+    /// If there is a problem starting the application, a [`LaunchError`] is
```

Comment:
Note that we now return a `LaunchError` if there's an error *after* launch as well. Happy to address this documentation with #1079 instead, which will change most of these signatures around again.

Reply:
Hmm...should we be returning a `LaunchError` if it's not necessarily related to launch? I don't believe Hyper's errors were directly exposed before — would it make sense to create an `enum` of the various error types and return that?